src/solovay_kitaev.py

```diff
@@ -0,0 +1,202 @@
+    
+    Returns:
+        Best approximating unitary from the basic gate set
+    """
+    best_dist = float('inf')
+    best_gate = None
+    
+    for gate in basic_gates:
+        dist = target.distance(gate)
+        if dist < best_dist:
+            best_dist = dist
+            best_gate = gate
+    
+    return best_gate
+
+def decompose_group_element(target: SU2Matrix, basic_gates: List[SU2Matrix], depth: int) -> Tuple[List[SU2Matrix], float]:
+    """
+    Decompose a target unitary into a sequence of basic gates using Solovay-Kitaev.
+    
+    Args:
+        target: Target unitary matrix to decompose
+        basic_gates: List of available basic gates
+        depth: Recursion depth for the algorithm
+    
+    Returns:
+        Tuple of (sequence of gates, approximation error)
+    """
+    if depth == 0:
+        best_approx = find_basic_approximation(target, basic_gates)
+        return [best_approx], target.distance(best_approx)
+    
+    # Recursive approximation
+    prev_sequence, prev_error = decompose_group_element(target, basic_gates, depth - 1)
+    
+    # If previous approximation is good enough, return it
+    # ERROR IS HARD CODED RIGHT NOW -> CHANGE THIS TO FIT USER-INPUT
+    if prev_error < 1e-6:
+        return prev_sequence, prev_error
+    
+    # Compute the error unitary
+    approx = prev_sequence[0]
+    for gate in prev_sequence[1:]:
+        approx = approx * gate
+    
+    error = target * approx.dagger()
+    
+    # Find Va and Vb such that their group commutator approximates the error
+    best_v = None
+    best_w = None
+    best_error = float('inf')
+    
+    for v in basic_gates:
+        for w in basic_gates:
+            comm = group_commutator(v, w)
+            curr_error = error.distance(comm)
+            if curr_error < best_error:
+                best_error = curr_error
+                best_v = v
+                best_w = w
+
+    # Construct the final sequence
+    result_sequence = []
+    result_sequence.extend(prev_sequence)
+    
+    # Add correction terms
+    if best_v is not None and best_w is not None:
+        v_sequence, error  = decompose_group_element(best_v, basic_gates, depth - 1)
+        w_sequence, error = decompose_group_element(best_w, basic_gates, depth - 1)
+
+        v_approx = v_sequence[0]
+        for gate in v_sequence[1:]:
+            v_approx = v_approx * gate
+        
+        w_approx = w_sequence[0]
+        for gate in w_sequence[1:]:
+            w_approx = w_approx * gate
+
+        result_sequence.extend([best_v, best_w, v_approx.dagger(), w_approx.dagger()])
+    
+    # Calculate final error
+    final_approx = result_sequence[0]
+    for gate in result_sequence[1:]:
+        final_approx = final_approx * gate
+    final_error = target.distance(final_approx)
+    
+    return result_sequence, final_error
+
+def solovay_kitaev(target: np.ndarray, basic_gates: List[np.ndarray], depth: int = 3) -> List[np.ndarray]:
```

Comment:
For a smoother interface, I believe `basis_gates` parameter should accept the [`BasisSet` enum](https://github.com/qBraid/pyqasm/blob/ef34299e3326ca5ef1d2bf24f101f17a87b6bb96/src/pyqasm/elements.py#L103) as a param and internally derive the basis gates from the [`BASIS_GATE_MAP`](https://github.com/qBraid/pyqasm/blob/ef34299e3326ca5ef1d2bf24f101f17a87b6bb96/src/pyqasm/maps/gates.py#L1090)